File: python/src/bursar/repositories/deduction.py
from __future__ import annotations

from bursar.repositories._types import DbQuery
from bursar.repositories._utils import validate_amount, validate_non_empty
from bursar.repositories.schemas import DeductionRow, DeductParams, RefundRow, RevokeRow


class DeductionRepository:
    """Repository for deduction, refund, and revocation operations.

    All methods call Postgres RPCs via the callproc function.
    Returns None when the RPC returns no rows (no result).
    Returns typed Pydantic models for successful results.
    """

    def __init__(self, callproc: DbQuery) -> None:
        self._callproc = callproc

    def deduct_with_allowance(self, params: DeductParams) -> DeductionRow | None:
        """Atomically deduct credits with allowance, cap, and floor checks.

        Args:
            params: DeductParams with user_id, amount, idempotency_key,
                min_balance, model, metadata, skip_allowance, period_start,
                and optional feature limit fields.

        Returns:
            DeductionRow if successful, None if the RPC returned no rows.
        """
        validate_non_empty(params.user_id, "user_id")
        validate_amount(params.amount, "amount")
        rows = self._callproc(
            "deduct_with_allowance",
            [
                params.user_id,
                params.amount,
                params.idempotency_key,
                params.min_balance,
                params.model,
                params.metadata,
                params.skip_allowance,
                params.period_start,
                params.feature,
                params.feature_max_calls,
                params.feature_action,
                params.feature_period_start,
                params.feature_period_end,
            ],
        )
        if not rows:
            return None
        return DeductionRow.model_validate(rows[0])

    def refund_credits(
        self,
        entry_id: str,
        amount: str | None,
        reason: str | None,
        metadata: str,
    ) -> RefundRow | None:
        """Refund a previous ledger entry, optionally for a partial amount.

        Args:
            entry_id: The original ledger entry ID to refund.
            amount: The amount to refund as a string, or None for full refund.
            reason: The refund reason, or None.
            metadata: JSON metadata string.

        Returns:
            RefundRow if successful, None if the RPC returned no rows.
        """
        validate_non_empty(entry_id, "entry_id")
        if amount is not None:
            validate_amount(amount, "amount")
        rows = self._callproc("refund_credits", [entry_id, amount, reason, metadata])
        if not rows:
            return None
        return RefundRow.model_validate(rows[0])

    def revoke_credits_by_entry_type(self, user_id: str, entry_type: str) -> RevokeRow | None:
        """Revoke credits for all transactions of a given type for a user.

        Args:
            user_id: The user ID.
            entry_type: The transaction type to revoke.

        Returns:
            RevokeRow with revocation details, or None if nothing to revoke.
        """
        validate_non_empty(user_id, "user_id")
        rows = self._callproc("revoke_credits_by_entry_type", [user_id, entry_type])
        if not rows:
            return None
        return RevokeRow.model_validate(rows[0])

File: javascript/src/repositories/deduction.ts
import { z } from "zod";
import type { CallProc } from "./types.js";
import { pgBoolean, safeParse } from "./_shared.js";

export const DeductionRowSchema = z
  .object({
    transaction_id: z.string().optional(),
    amount: z
      .union([z.string(), z.number()] as const)
      .nullable()
      .optional(),
    balance_after: z
      .union([z.string(), z.number()] as const)
      .nullable()
      .optional(),
    allowance_consumed: z
      .union([z.string(), z.number()] as const)
      .nullable()
      .optional(),
    idempotent: pgBoolean.nullable().optional(),
    cap_warning: z.string().nullable().optional(),
    feature_limit_warning: z.string().nullable().optional(),
    bucket_breakdown: z
      .record(z.string(), z.union([z.string(), z.number()] as const))
      .nullable()
      .optional(),
    error: z.string().optional(),
    user_id: z.string().optional(),
  })
  .passthrough();

export const RefundRowSchema = z
  .object({
    refund_transaction_id: z.string().optional(),
    user_id: z.string().optional(),
    amount: z
      .union([z.string(), z.number()] as const)
      .nullable()
      .optional(),
    new_balance: z
      .union([z.string(), z.number()] as const)
      .nullable()
      .optional(),
    bucket_breakdown: z
      .record(z.string(), z.union([z.string(), z.number()] as const))
      .nullable()
      .optional(),
    error: z.string().optional(),
  })
  .passthrough();

export const RevokeRowSchema = z
  .object({
    user_id: z.string().optional(),
    amount: z
      .union([z.string(), z.number()] as const)
      .nullable()
      .optional(),
    new_balance: z
      .union([z.string(), z.number()] as const)
      .nullable()
      .optional(),
    bucket: z.string().nullable().optional(),
  })
  .passthrough();

export type DeductionRow = z.infer<typeof DeductionRowSchema>;
export type RefundRow = z.infer<typeof RefundRowSchema>;
export type RevokeRow = z.infer<typeof RevokeRowSchema>;

/** Typed parameters for deductWithAllowance, replacing 13 positional parameters. */
export interface DeductParams {
  /** The user ID to deduct from. */
  userId: string;
  /** Amount to deduct as a decimal string. */
  amount: string;
  /** Idempotency key for replay safety. */
  idempotencyKey: string | null;
  /** Minimum balance floor after deduction. */
  minBalance: string;
  /** Model identifier for rate tracking. */
  model: string | null;
  /** JSON-encoded metadata. */
  metadata: string;
  /** Skip free-allowance consumption. */
  skipAllowance: boolean;
  /** Free-allowance window start override. */
  periodStart: string | null;
  /** Feature name for per-feature limits. */
  feature: string | null;
  /** Maximum calls for feature limit enforcement. */
  featureMaxCalls: number | null;
  /** Action when feature limit exceeded: deny, warn, notify. */
  featureOnExceed: string | null;
  /** Feature limit window start. */
  featurePeriodStart: string | null;
  /** Feature limit window end. */
  featurePeriodEnd: string | null;
}

/** Repository for credit deduction operations. */
export class DeductionRepository {
  constructor(private callproc: CallProc) {}

  /** Atomically deduct credits with allowance consumption, spend cap enforcement, and floor check. */
  async deductWithAllowance(params: DeductParams): Promise<DeductionRow> {
    const rows = await this.callproc("deduct_with_allowance", [
      params.userId,
      params.amount,
      params.idempotencyKey,
      params.minBalance,
      params.model,
      params.metadata,
      params.skipAllowance,
      params.periodStart,
      params.feature,
      params.featureMaxCalls,
      params.featureOnExceed,
      params.featurePeriodStart,
      params.featurePeriodEnd,
    ]);
    return safeParse(
      DeductionRowSchema,
      rows?.[0] ?? {},
      "DeductionRepository.deductWithAllowance",
    );
  }

  /** Refund a previous credit deduction. */
  async refundCredits(
    transactionId: string,
    amount: string | null,
    reason: string | null,
    metadata: string,
  ): Promise<RefundRow> {
    const rows = await this.callproc("refund_credits", [transactionId, amount, reason, metadata]);
    return safeParse(RefundRowSchema, rows?.[0] ?? {}, "DeductionRepository.refundCredits");
  }

  /** Revoke credits by transaction type. */
  async revokeCreditsByTxType(userId: string, txType: string): Promise<RevokeRow> {
    const rows = await this.callproc("revoke_credits_by_tx_type", [userId, txType]);
    return safeParse(RevokeRowSchema, rows?.[0] ?? {}, "DeductionRepository.revokeCreditsByTxType");
  }
}
